sdks/java/core/src/main/java/org/apache/beam/sdk/options/ValueProvider.java

```diff
@@ -0,0 +1,226 @@
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class RuntimeValueProvider<T> implements ValueProvider<T>, Serializable {
+    private static ConcurrentHashMap<Long, PipelineOptions> optionsMap =
+      new ConcurrentHashMap<>();
+
+    private final Class<? extends PipelineOptions> klass;
+    private final String methodName;
+    private final T defaultValue;
+    private final Long optionsId;
+
+    /**
+     * Creates a {@link RuntimeValueProvider} that will query the provided
+     * optionsId for a value.
+     */
+    RuntimeValueProvider(String methodName, Class<? extends PipelineOptions> klass,
+                         Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = null;
+      this.optionsId = optionsId;
+    }
+
+    /**
+     * Creates a {@link RuntimeValueProvider} that will query the provided
+     * optionsId for a value, or use the default if no value is available.
+     */
+    RuntimeValueProvider(String methodName, Class<? extends PipelineOptions> klass,
+      T defaultValue, Long optionsId) {
+      this.methodName = methodName;
+      this.klass = klass;
+      this.defaultValue = defaultValue;
+      this.optionsId = optionsId;
+    }
+
+    static void setRuntimeOptions(PipelineOptions runtimeOptions) {
+      optionsMap.put(runtimeOptions.getOptionsId(), runtimeOptions);
+    }
+
+    @Override
+    public T get() {
+      PipelineOptions options = optionsMap.get(optionsId);
+      if (options == null) {
+        if (defaultValue != null) {
+          return defaultValue;
+        }
+        throw new RuntimeException("Not called from a runtime context.");
+      }
+      try {
+        Method method = klass.getMethod(methodName);
+        PipelineOptions methodOptions = options.as(klass);
+        InvocationHandler handler = Proxy.getInvocationHandler(methodOptions);
+        return ((ValueProvider<T>) handler.invoke(methodOptions, method, null)).get();
+      } catch (Throwable e) {
+        throw new RuntimeException("Unable to load runtime value.", e);
+      }
+    }
+
+    @Override
+    public boolean shouldValidate() {
+      return defaultValue != null;
+    }
+  }
+
+  /**
+   * Serializer for {@link ValueProvider}.
+   */
+  static class Serializer extends JsonSerializer<ValueProvider<?>> {
+    @Override
+    public void serialize(ValueProvider<?> value, JsonGenerator jgen,
+                          SerializerProvider provider) throws IOException {
+      if (value.shouldValidate()) {
+        jgen.writeObject(value.get());
+      } else {
+        jgen.writeNull();
+      }
+    }
+  }
+
+  /**
+   * Deserializer for {@link ValueProvider}, which handles type marshalling.
+   */
+  static class Deserializer extends JsonDeserializer<ValueProvider<?>>
+    implements ContextualDeserializer {
+
+    private final JavaType innerType;
+
+    // A 0-arg constructor is required.
```

Comment:
`javac` only requires that a constructor exist. Since you have the `Deserializer(JavaType innerType)`, this cannot quite be right.